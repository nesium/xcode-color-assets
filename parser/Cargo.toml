[package]
name = "parser"
version = "0.1.0"
authors = ["mb <mb@nesium.com>"]
edition = "2018"

[dependencies]
nom = "5.0"

[dev-dependencies]
insta = "0.10"