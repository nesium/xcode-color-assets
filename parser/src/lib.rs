pub mod ast;
mod parser;

pub use self::parser::{parse_document, parse_document_from_file};
