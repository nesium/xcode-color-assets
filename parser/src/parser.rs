use super::ast::{
  Color, ColorSet, ColorSetValue, Declaration, Document, DocumentItem, RuleSet, RuleSetItem, Value,
};
use nom::branch::alt;
use nom::bytes::complete::{tag, take_while_m_n};
use nom::character::complete::{alphanumeric1, char, digit1, multispace0, newline, space0, space1};
use nom::combinator::{all_consuming, cut, map, map_res, opt};
use nom::error::{context, convert_error, ParseError, VerboseError};
use nom::multi::{many0, separated_list};
use nom::number::complete::float;
use nom::sequence::{delimited, preceded, separated_pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use std::fmt;
use std::fs::File;
use std::io::prelude::Read;
use std::str;

#[derive(Debug)]
pub struct Error {
  error: String,
}

impl std::error::Error for Error {
  fn description(&self) -> &str {
    &self.error
  }
}

impl fmt::Display for Error {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "Error: {}", self.error)
  }
}

fn parse_hex_value(input: &str) -> Result<u32, std::num::ParseIntError> {
  u32::from_str_radix(input, 16)
}

fn colorset_from_declarations(
  decl1: Declaration<ColorSetValue>,
  decl2: Declaration<ColorSetValue>,
) -> Result<ColorSet, Error> {
  match (decl1.identifier.as_str(), decl2.identifier.as_str()) {
    ("light", "dark") => Ok(ColorSet {
      light: decl1.value,
      dark: decl2.value,
    }),
    ("dark", "light") => Ok(ColorSet {
      light: decl2.value,
      dark: decl1.value,
    }),
    _ => Err(Error {
      error: format!(
        "Expected light & dark properties. Found {}, {}.",
        decl1.identifier, decl2.identifier
      ),
    }),
  }
}

fn line_delimiter<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
  cut(terminated(preceded(space0, newline), multispace0))(input)
}

pub fn parse_document(input: String) -> Result<Document, Error> {
  let mut modified_input = input;
  modified_input.push_str("\n");

  let result: IResult<&str, Document, VerboseError<&str>> = map(
    all_consuming(delimited(
      multispace0,
      separated_list(
        line_delimiter,
        alt((
          map(variable, DocumentItem::Variable),
          map(ruleset, DocumentItem::RuleSet),
          map(declaration(value), DocumentItem::Declaration),
        )),
      ),
      multispace0,
    )),
    |res| Document { items: res },
  )(&modified_input);

  match result {
    Ok((_, doc)) => Ok(doc),
    Err(Err::Error(e)) | Err(Err::Failure(e)) => Err(Error {
      error: convert_error(&modified_input, e).replace("'\n'", "'\\n'"),
    }),
    _ => {
      eprintln!("An unknown error occurred.");
      std::process::exit(0x0100)
    }
  }
}

pub fn parse_document_from_file(filepath: &str) -> Result<Document, Error> {
  let mut f = File::open(filepath).unwrap();
  let mut buffer = vec![];
  f.read_to_end(&mut buffer).unwrap();

  let contents = match str::from_utf8(&buffer) {
    Ok(v) => v,
    Err(e) => panic!(
      "Could not read contents of file {}. Reason: {}",
      filepath, e
    ),
  };

  parse_document(contents.to_string())
}

pub fn ruleset<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, RuleSet, E> {
  let body = |input| {
    delimited(
      terminated(char('{'), multispace0),
      cut(many0(terminated(
        alt((
          map(ruleset, RuleSetItem::RuleSet),
          map(declaration(value), RuleSetItem::Declaration),
        )),
        line_delimiter,
      ))),
      preceded(multispace0, char('}')),
    )(input)
  };

  context(
    "RuleSet",
    map(separated_pair(identifier, space0, body), |res| RuleSet {
      identifier: res.0.to_string(),
      items: res.1,
    }),
  )(input)
}

pub fn variable<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, Declaration<Value>, E> {
  context(
    "Variable",
    map(
      separated_pair(
        variable_identifier,
        cut(terminated(preceded(space0, char(':')), space0)),
        value,
      ),
      |res| {
        let (identifier, value) = res;
        Declaration { identifier, value }
      },
    ),
  )(input)
}

fn value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Value, E> {
  preceded(
    space0,
    alt((
      map(colorset_value, |res| res.into()),
      map(colorset, Value::ColorSet),
    )),
  )(input)
}

fn colorset_value<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, ColorSetValue, E> {
  preceded(
    space0,
    alt((
      map(hex_color, ColorSetValue::Color),
      map(rgba_color, ColorSetValue::Color),
      map(variable_identifier, ColorSetValue::Variable),
    )),
  )(input)
}

fn rgba_color<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Color, E> {
  let u8_value = move |input: &'a str| {
    preceded(
      space0,
      cut(context(
        "Value between 0 - 255",
        map_res(digit1, |s: &str| s.parse::<u8>()),
      )),
    )(input)
  };
  let delimiter = move |input: &'a str| preceded(space0, cut(char(',')))(input);

  context(
    "RGBA Value",
    map(
      preceded(
        tag("rgba"),
        delimited(
          preceded(space0, char('(')),
          tuple((
            terminated(u8_value, delimiter),
            terminated(u8_value, delimiter),
            terminated(u8_value, delimiter),
            preceded(space0, float),
          )),
          preceded(space0, char(')')),
        ),
      ),
      |res| {
        let (r, g, b, a) = res;
        Color { r, g, b, a }
      },
    ),
  )(input)
}

fn hex_color<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Color, E> {
  map(
    tuple((
      char('#'),
      cut(hex_value),
      opt(preceded(space1, alpha_value)),
    )),
    |res| {
      let (_, rgb, alpha) = res;
      let r = ((rgb >> 16) & 0xff) as u8;
      let g = ((rgb >> 8) & 0xff) as u8;
      let b = (rgb & 0xff) as u8;
      let a = alpha.unwrap_or(1.0);
      Color { r, g, b, a }
    },
  )(input)
}

pub fn colorset<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, ColorSet, E> {
  context(
    "ColorSet",
    map_res(
      delimited(
        cut(char('(')),
        separated_pair(
          cut(declaration(colorset_value)),
          cut(preceded(space0, char(','))),
          cut(declaration(colorset_value)),
        ),
        cut(char(')')),
      ),
      |res| colorset_from_declarations(res.0, res.1),
    ),
  )(input)
}

pub fn identifier<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, String, E> {
  context(
    "Identifier",
    map(preceded(space0, alphanumeric1), |ident: &str| {
      ident.to_string()
    }),
  )(input)
}

pub fn variable_identifier<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, String, E> {
  context(
    "Variable Identifier",
    map(
      preceded(space0, preceded(char('$'), alphanumeric1)),
      |ident: &str| ident.to_string(),
    ),
  )(input)
}

pub fn declaration<'a, F, V, E: ParseError<&'a str>>(
  value: F,
) -> impl Fn(&'a str) -> IResult<&'a str, Declaration<V>, E>
where
  F: Fn(&'a str) -> IResult<&'a str, V, E>,
{
  move |input: &'a str| {
    context(
      "Declaration",
      map(
        separated_pair(identifier, cut(preceded(space0, char(':'))), &value),
        |res| {
          let (identifier, value) = res;
          Declaration { identifier, value }
        },
      ),
    )(input)
  }
}

fn hex_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, u32, E> {
  context(
    "Hex Value",
    map_res(take_while_m_n(6, 6, is_hex_digit), parse_hex_value),
  )(input)
}

fn alpha_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, f32, E> {
  context(
    "Alpha Value",
    map_res(terminated(digit1, tag("%")), |input| {
      u32::from_str_radix(input, 10).map(|val| (val as f32) / 100.0)
    }),
  )(input)
}

fn is_hex_digit(c: char) -> bool {
  c.is_digit(16)
}
