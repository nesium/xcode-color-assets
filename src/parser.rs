use super::ast::{
  Color, ColorSet, ColorSetValue, Declaration, Document, DocumentItem, RuleSet, RuleSetItem, Value,
};
use nom::branch::alt;
use nom::bytes::complete::{tag, take_while_m_n};
use nom::character::complete::{alphanumeric1, char, digit1, multispace0, newline, space0, space1};
use nom::combinator::{map, map_res, opt};
use nom::error::{context, ParseError, VerboseError};
use nom::multi::{many0, separated_list};
use nom::sequence::{preceded, separated_pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use std::fs::File;
use std::io::prelude::Read;
use std::str;

#[allow(dead_code)]
fn parse_hex_value(input: &str) -> Result<u32, std::num::ParseIntError> {
  u32::from_str_radix(input, 16)
}

#[allow(dead_code)]
fn colorset_from_declarations(
  decl1: Declaration<ColorSetValue>,
  decl2: Declaration<ColorSetValue>,
) -> Result<ColorSet, String> {
  match (decl1.identifier.as_str(), decl2.identifier.as_str()) {
    ("light", "dark") => Ok(ColorSet {
      light: decl1.value,
      dark: decl2.value,
    }),
    ("dark", "light") => Ok(ColorSet {
      light: decl2.value,
      dark: decl1.value,
    }),
    _ => Err(format!(
      "Expected light & dark properties. Found {}, {}.",
      decl1.identifier, decl2.identifier
    )),
  }
}

#[allow(dead_code)]
fn line_delimiter(input: &str) -> IResult<&str, char> {
  terminated(preceded(space0, newline), multispace0)(input)
}

#[allow(dead_code)]
pub fn parse_document(input: &str) -> IResult<&str, Document> {
  map(
    preceded(
      multispace0,
      separated_list(
        line_delimiter,
        alt((
          map(variable, DocumentItem::Variable),
          map(ruleset, DocumentItem::RuleSet),
        )),
      ),
    ),
    |res| Document { items: res },
  )(input)
}

pub fn parse_document_from_file(filepath: &str) -> Result<Document, VerboseError<String>> {
  let mut f = File::open(filepath).unwrap();
  let mut buffer = vec![];
  f.read_to_end(&mut buffer).unwrap();

  let contents = match str::from_utf8(&buffer) {
    Ok(v) => v,
    Err(e) => panic!(
      "Could not read contents of file {}. Reason: {}",
      filepath, e
    ),
  };

  match parse_document(&contents) {
    Ok(res) => Ok(res.1),
    Err(Err::Error(e)) | Err(Err::Failure(e)) => {
      Err(VerboseError::from_error_kind(contents.to_string(), e.1))
    }
    _ => {
      eprintln!("Something went wrong");
      std::process::exit(0x0100)
    }
  }
}

#[allow(dead_code)]
pub fn ruleset(input: &str) -> IResult<&str, RuleSet> {
  let body = |input| {
    preceded(
      terminated(char('{'), multispace0),
      terminated(
        many0(alt((
          map(terminated(ruleset, line_delimiter), RuleSetItem::RuleSet),
          map(
            terminated(declaration(value), line_delimiter),
            RuleSetItem::Declaration,
          ),
        ))),
        preceded(multispace0, char('}')),
      ),
    )(input)
  };

  context(
    "ruleset",
    map(tuple((identifier, space0, body)), |res| RuleSet {
      identifier: res.0.to_string(),
      items: res.2,
    }),
  )(input)
}

pub fn variable(input: &str) -> IResult<&str, Declaration<Value>> {
  map(
    separated_pair(
      identifier,
      terminated(preceded(space0, char('=')), space0),
      value,
    ),
    |res| {
      let (identifier, value) = res;
      Declaration { identifier, value }
    },
  )(input)
}

fn value(input: &str) -> IResult<&str, Value> {
  alt((
    map(colorset_value, |res| res.into()),
    map(colorset, Value::ColorSet),
  ))(input)
}

fn colorset_value(input: &str) -> IResult<&str, ColorSetValue> {
  alt((
    map(hex_color, ColorSetValue::Color),
    map(preceded(char('$'), identifier), ColorSetValue::Variable),
  ))(input)
}

fn hex_color(input: &str) -> IResult<&str, Color> {
  map(
    tuple((tag("#"), hex_value, opt(preceded(space1, alpha_value)))),
    |res| {
      let (_, rgb, alpha) = res;
      let r = ((rgb >> 16) & 0xff) as u8;
      let g = ((rgb >> 8) & 0xff) as u8;
      let b = (rgb & 0xff) as u8;
      let a = alpha.unwrap_or(1.0);
      Color { r, g, b, a }
    },
  )(input)
}

pub fn colorset(input: &str) -> IResult<&str, ColorSet> {
  map_res(
    tuple((
      tag("("),
      declaration(colorset_value),
      space0,
      tag(","),
      multispace0,
      declaration(colorset_value),
      tag(")"),
    )),
    |res| colorset_from_declarations(res.1, res.5),
  )(input)
}

pub fn identifier(input: &str) -> IResult<&str, String> {
  map(preceded(space0, alphanumeric1), |ident: &str| {
    ident.to_string()
  })(input)
}

pub fn declaration<F, V>(value: F) -> impl Fn(&str) -> IResult<&str, Declaration<V>>
where
  F: Fn(&str) -> IResult<&str, V>,
{
  move |input: &str| {
    map(
      separated_pair(
        identifier,
        terminated(preceded(space0, char(':')), space0),
        &value,
      ),
      |res| {
        let (identifier, value) = res;
        Declaration { identifier, value }
      },
    )(input)
  }
}

fn hex_value(input: &str) -> IResult<&str, u32> {
  map_res(take_while_m_n(6, 6, is_hex_digit), parse_hex_value)(input)
}

fn alpha_value(input: &str) -> IResult<&str, f32> {
  map_res(terminated(digit1, tag("%")), |input| {
    u32::from_str_radix(input, 10).map(|val| (val as f32) / 100.0)
  })(input)
}

fn is_hex_digit(c: char) -> bool {
  c.is_digit(16)
}
