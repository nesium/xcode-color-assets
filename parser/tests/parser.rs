use insta::{assert_debug_snapshot_matches, assert_snapshot_matches};
use parser::parse_document;

#[test]
fn empty_document() {
  test_parse_doc("empty_document", "");
}

#[test]
fn variable_only() {
  test_parse_doc("variable_only", "$var: #ff00ff");
}

#[test]
fn declaration_only() {
  test_parse_doc("declaration_only", "a: #00ff00");
}

#[test]
fn empty_ruleset() {
  test_parse_doc("empty_ruleset_1", "a {}");
  test_parse_doc(
    "empty_ruleset_2",
    r#"
    a { b {}
    }
  "#,
  );
}

#[test]
fn variable_declaration_ruleset_order() {
  test_parse_doc(
    "variable_declaration_ruleset_order_1",
    r#"
      a: $var
      a {}
      $var: #ff00ff
  "#,
  );
  test_parse_doc(
    "variable_declaration_ruleset_order_2",
    r#"
      $var: #ff00ff
      a {}
      a: $var
  "#,
  );
}

#[test]
fn garbage_doc() {
  test_parse_doc("garbage_doc_1", "something");
  test_parse_doc("garbage_doc_2", ".");
  test_parse_doc("garbage_doc_3", "{}");
  test_parse_doc("garbage_doc_4", "$");
}

#[test]
fn color_variables() {
  test_parse_doc("color_variables_1", "$myColor : #ff00ff 44%");
  test_parse_doc("color_variables_2", "$a :#4224be");
  test_parse_doc("color_variables_3", "$a1: #4B0FC6");
  test_parse_doc("color_variables_4", "$a1: $a0");
  test_parse_doc("color_variables_5", "$a: rgba(0, 127, 255, 0.5)");
  test_parse_doc("color_variables_6", "$a: rgba (0, 127, 255, 0.5)");
  test_parse_doc("color_variables_7", "$a: (light: #ff00ff, dark: #00ff00)");
}

#[test]
fn garbage_hex_value() {
  test_parse_doc("garbage_hex_value_1", "a: #ff0zff");
  test_parse_doc("garbage_hex_value_2", "a: #fff");
  test_parse_doc("garbage_hex_value_3", "a: #ffffff 33?");
  test_parse_doc("garbage_hex_value_4", "a: #fffffff");
}

#[test]
fn garbage_rgba_values() {
  test_parse_doc("garbage_rgba_value_1", "$a: rgba(0, 127; 255, 0.5)");
  test_parse_doc("garbage_rgba_value_2", "$a: rgba0, 127, 255, 0.5)");
  test_parse_doc("garbage_rgba_value_3", "$a: rgba(0, 300, 255, 0.5)")
}

#[test]
fn color_sets() {
  test_parse_doc(
    "color_sets",
    r#"
    a: (light: #ff00ff 30%, dark: #00ff00)
    b: (dark: #00ff00, light: #ff00ff 30%)
    c: (dark: $applicationBackgroundLight, light: #ff00ff 30%)
    d: (light: rgba(33, 199, 201, 1), dark: #ff00ff 30%)
  "#,
  )
}

#[test]
fn typos_in_color_sets() {
  test_parse_doc(
    "typos_in_color_sets_1",
    r#"
    a: (light = #ff00ff 30%, dark: #00ff00)
  "#,
  );
  test_parse_doc(
    "typos_in_color_sets_2",
    r#"
    a: (light: #ff00ff 30%; dark: #00ff00)
  "#,
  );
  test_parse_doc(
    "typos_in_color_sets_3",
    r#"
    a: (light: #ff00ff 30%, dark: variable)
  "#,
  );
  test_parse_doc(
    "typos_in_color_sets_4",
    r#"
    a: [light: #ff00ff 30%, dark: #00ff00)
  "#,
  );
  test_parse_doc(
    "typos_in_color_sets_5",
    r#"
    a: (light: #ff00ff 30%, dark: #00ff00]
  "#,
  );
  test_parse_doc(
    "typos_in_color_sets_6",
    r#"
    a: (light: #ff00ff 30%, dark: #00ff00)
    b: (dark: #00ff00, light: #ff00ff 30%)
    c: (dark: .applicationBackgroundLight, light: #ff00ff 30%)
  "#,
  )
}

#[test]
fn document() {
  test_parse_doc(
    "document",
    r#"
    $mediumBright: #aabbcc 33%
    $red: #ff0000
    $white: #ffffff

    ApplicationBackground: (light: $white, dark: #141517)

    Text {
      Primary: (light: #151618, dark: #E7E8EA)
      Secondary: (light: #75767A, dark: #85868A)
    }

    NumericInput {
      ActionKey {
        Background: (light: $red, dark: #ff00ff)
        Highlight: (light: #cccccc, dark: #000000)
      }
    }

    Background: (light: #D6D9DE, dark: #313131)
  "#,
  );
}

fn test_parse_doc(test_name: &str, contents: &str) {
  match parse_document(contents.to_string()) {
    Ok(doc) => assert_debug_snapshot_matches!(test_name, doc),
    Err(e) => assert_snapshot_matches!(test_name, format!("{}", e)),
  }
}
