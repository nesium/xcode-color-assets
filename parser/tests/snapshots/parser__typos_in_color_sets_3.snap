---
created: "2019-08-18T08:19:52.014951Z"
creator: insta@0.10.1
source: parser/tests/parser.rs
expression: "format!(\"{}\" , e)"
---
Error: 0: at line 1:
    a: (light: #ff00ff 30%, dark: variable)
                                  ^
expected '$', found v

1: at line 1, in Variable Identifier:
    a: (light: #ff00ff 30%, dark: variable)
                                  ^

2: at line 1, in Alt:
    a: (light: #ff00ff 30%, dark: variable)
                                  ^

3: at line 1, in Declaration:
    a: (light: #ff00ff 30%, dark: variable)
                           ^

4: at line 1, in ColorSet:
    a: (light: #ff00ff 30%, dark: variable)
       ^

5: at line 1, in Declaration:
    a: (light: #ff00ff 30%, dark: variable)
    ^


